models/MortgageDetail.go

```diff
@@ -23,8 +26,121 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
+	var l = logger.WithFields(logrus.Fields{
+		"method":              "mortgageStocksAction",
+		"param_userId":        user.Id,
+		"param_stockId":       stockID,
+		"param_stockQuantity": stockQuantity,
+		"param_mortgagePrice": mortgagePrice,
+	})
+
+	stockOwned, err := getSingleStockCount(user, stockID)
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	if stockQuantity > stockOwned {
+		l.Errorf("Insufficient stocks to mortgage. Have %d, want %d", stockOwned, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	stocksInBank, err := getStocksInBank(user.Id, stockID, mortgagePrice, tx)
+	if (err != nil && err != InvalidRetrievePriceError{}) { /* This type of error not applicable here */
+		l.Error(err)
+		return 0, err
+	}
+
+	if stocksInBank == 0 {
+		sql := "INSERT into MortgageDetails (userId, stockId, stocksInBank, mortgagePrice) VALUES (?, ?, ?, ?)"
+		err = tx.Exec(sql, user.Id, stockID, stockQuantity, mortgagePrice).Error
+	} else {
+		sql := "UPDATE MortgageDetails SET stocksInBank=? WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, stocksInBank+stockQuantity, user.Id, stockID, mortgagePrice).Error
+	}
+
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	return int64(mortgagePrice) * stockQuantity * MORTGAGE_DEPOSIT_RATE / 100, nil
+}
+
+// retrieveStocksAction returns total transaction amount and stock quantity which could be retrieved
+func retrieveStocksAction(userID, stockID uint32, stockQuantity int64, userCash, retrievePrice uint64, tx *gorm.DB) (int64, error) {
+	var l = logger.WithFields(logrus.Fields{
+		"method":              "retrieveStocksAction",
+		"param_userId":        userID,
+		"param_stockId":       stockID,
+		"param_stockQuantity": stockQuantity,
+		"param_retrievePrice": retrievePrice,
+	})
+
+	l.Debugf("Retrieving stocks in action")
+
+	stocksInBank, err := getStocksInBank(userID, stockID, retrievePrice, tx)
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	if int64(userCash/retrievePrice) < stockQuantity {
+		l.Errorf("Insufficient cash with user. Have %d, want %d", userCash, stockQuantity*int64(retrievePrice))
+		return 0, NotEnoughCashError{}
+	}
+
+	if stockQuantity < stocksInBank {
+		sql := "UPDATE MortgageDetails SET stocksInBank=? WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, stocksInBank-stockQuantity, userID, stockID, retrievePrice).Error
+
+	} else if stockQuantity == stocksInBank { /* So we can delete that entire row */
+		sql := "DELETE from MortgageDetails WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, userID, stockID, retrievePrice).Error
+
+	} else { /* stockQuantity to retrieve > stocksInBank */
+		l.Errorf("Insufficient stocks in mortgage. Have %d, want %d", stocksInBank, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	return -int64(retrievePrice) * stockQuantity * MORTGAGE_RETRIEVE_RATE / 100, nil
```

Comment:
So say a stock is worth Rs. 100.
If I mortgage a stock, I get Rs. 80 (80%) back and when I buy the stock back I pay Rs. 90 (90%)?

Reply:
Yes

Reply:
You are right